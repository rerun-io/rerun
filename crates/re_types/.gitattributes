# DO NOT EDIT! This file is generated by crates/re_types_builder/src/lib.rs

.gitattributes linguist-generated=true
src/archetypes/annotation_context.rs linguist-generated=true
src/archetypes/arrows3d.rs linguist-generated=true
src/archetypes/asset3d.rs linguist-generated=true
src/archetypes/bar_chart.rs linguist-generated=true
src/archetypes/boxes2d.rs linguist-generated=true
src/archetypes/boxes3d.rs linguist-generated=true
src/archetypes/clear.rs linguist-generated=true
src/archetypes/depth_image.rs linguist-generated=true
src/archetypes/disconnected_space.rs linguist-generated=true
src/archetypes/image.rs linguist-generated=true
src/archetypes/line_strips2d.rs linguist-generated=true
src/archetypes/line_strips3d.rs linguist-generated=true
src/archetypes/mesh3d.rs linguist-generated=true
src/archetypes/mod.rs linguist-generated=true
src/archetypes/pinhole.rs linguist-generated=true
src/archetypes/points2d.rs linguist-generated=true
src/archetypes/points3d.rs linguist-generated=true
src/archetypes/segmentation_image.rs linguist-generated=true
src/archetypes/tensor.rs linguist-generated=true
src/archetypes/text_document.rs linguist-generated=true
src/archetypes/text_log.rs linguist-generated=true
src/archetypes/time_series_scalar.rs linguist-generated=true
src/archetypes/transform3d.rs linguist-generated=true
src/archetypes/view_coordinates.rs linguist-generated=true
src/components/annotation_context.rs linguist-generated=true
src/components/blob.rs linguist-generated=true
src/components/class_id.rs linguist-generated=true
src/components/clear_is_recursive.rs linguist-generated=true
src/components/color.rs linguist-generated=true
src/components/depth_meter.rs linguist-generated=true
src/components/disconnected_space.rs linguist-generated=true
src/components/draw_order.rs linguist-generated=true
src/components/half_sizes2d.rs linguist-generated=true
src/components/half_sizes3d.rs linguist-generated=true
src/components/instance_key.rs linguist-generated=true
src/components/keypoint_id.rs linguist-generated=true
src/components/line_strip2d.rs linguist-generated=true
src/components/line_strip3d.rs linguist-generated=true
src/components/material.rs linguist-generated=true
src/components/media_type.rs linguist-generated=true
src/components/mesh_properties.rs linguist-generated=true
src/components/mod.rs linguist-generated=true
src/components/origin2d.rs linguist-generated=true
src/components/origin3d.rs linguist-generated=true
src/components/out_of_tree_transform3d.rs linguist-generated=true
src/components/pinhole_projection.rs linguist-generated=true
src/components/position2d.rs linguist-generated=true
src/components/position3d.rs linguist-generated=true
src/components/radius.rs linguist-generated=true
src/components/resolution.rs linguist-generated=true
src/components/rotation3d.rs linguist-generated=true
src/components/scalar.rs linguist-generated=true
src/components/scalar_scattering.rs linguist-generated=true
src/components/tensor_data.rs linguist-generated=true
src/components/text.rs linguist-generated=true
src/components/text_log_level.rs linguist-generated=true
src/components/transform3d.rs linguist-generated=true
src/components/vector3d.rs linguist-generated=true
src/components/view_coordinates.rs linguist-generated=true
src/datatypes/angle.rs linguist-generated=true
src/datatypes/annotation_info.rs linguist-generated=true
src/datatypes/class_description.rs linguist-generated=true
src/datatypes/class_description_map_elem.rs linguist-generated=true
src/datatypes/class_id.rs linguist-generated=true
src/datatypes/color.rs linguist-generated=true
src/datatypes/float32.rs linguist-generated=true
src/datatypes/keypoint_id.rs linguist-generated=true
src/datatypes/keypoint_pair.rs linguist-generated=true
src/datatypes/mat3x3.rs linguist-generated=true
src/datatypes/mat4x4.rs linguist-generated=true
src/datatypes/material.rs linguist-generated=true
src/datatypes/mesh_properties.rs linguist-generated=true
src/datatypes/mod.rs linguist-generated=true
src/datatypes/quaternion.rs linguist-generated=true
src/datatypes/rotation3d.rs linguist-generated=true
src/datatypes/rotation_axis_angle.rs linguist-generated=true
src/datatypes/scale3d.rs linguist-generated=true
src/datatypes/tensor_buffer.rs linguist-generated=true
src/datatypes/tensor_data.rs linguist-generated=true
src/datatypes/tensor_dimension.rs linguist-generated=true
src/datatypes/transform3d.rs linguist-generated=true
src/datatypes/translation_and_mat3x3.rs linguist-generated=true
src/datatypes/translation_rotation_scale3d.rs linguist-generated=true
src/datatypes/utf8.rs linguist-generated=true
src/datatypes/uvec2d.rs linguist-generated=true
src/datatypes/uvec3d.rs linguist-generated=true
src/datatypes/uvec4d.rs linguist-generated=true
src/datatypes/vec2d.rs linguist-generated=true
src/datatypes/vec3d.rs linguist-generated=true
src/datatypes/vec4d.rs linguist-generated=true
src/testing/archetypes/affix_fuzzer1.rs linguist-generated=true
src/testing/archetypes/mod.rs linguist-generated=true
src/testing/components/affix_fuzzer1.rs linguist-generated=true
src/testing/components/affix_fuzzer10.rs linguist-generated=true
src/testing/components/affix_fuzzer11.rs linguist-generated=true
src/testing/components/affix_fuzzer12.rs linguist-generated=true
src/testing/components/affix_fuzzer13.rs linguist-generated=true
src/testing/components/affix_fuzzer14.rs linguist-generated=true
src/testing/components/affix_fuzzer15.rs linguist-generated=true
src/testing/components/affix_fuzzer16.rs linguist-generated=true
src/testing/components/affix_fuzzer17.rs linguist-generated=true
src/testing/components/affix_fuzzer18.rs linguist-generated=true
src/testing/components/affix_fuzzer19.rs linguist-generated=true
src/testing/components/affix_fuzzer2.rs linguist-generated=true
src/testing/components/affix_fuzzer20.rs linguist-generated=true
src/testing/components/affix_fuzzer21.rs linguist-generated=true
src/testing/components/affix_fuzzer3.rs linguist-generated=true
src/testing/components/affix_fuzzer4.rs linguist-generated=true
src/testing/components/affix_fuzzer5.rs linguist-generated=true
src/testing/components/affix_fuzzer6.rs linguist-generated=true
src/testing/components/affix_fuzzer7.rs linguist-generated=true
src/testing/components/affix_fuzzer8.rs linguist-generated=true
src/testing/components/affix_fuzzer9.rs linguist-generated=true
src/testing/components/mod.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer1.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer2.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer20.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer21.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer3.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer4.rs linguist-generated=true
src/testing/datatypes/affix_fuzzer5.rs linguist-generated=true
src/testing/datatypes/flattened_scalar.rs linguist-generated=true
src/testing/datatypes/mod.rs linguist-generated=true
src/testing/datatypes/primitive_component.rs linguist-generated=true
src/testing/datatypes/string_component.rs linguist-generated=true
