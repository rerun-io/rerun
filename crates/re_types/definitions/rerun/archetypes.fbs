include "./archetypes/annotation_context.fbs";
include "./archetypes/arrows3d.fbs";
include "./archetypes/asset3d.fbs";
include "./archetypes/bar_chart.fbs";
include "./archetypes/boxes2d.fbs";
include "./archetypes/boxes3d.fbs";
include "./archetypes/clear.fbs";
include "./archetypes/depth_image.fbs";
include "./archetypes/disconnected_space.fbs";
include "./archetypes/image.fbs";
include "./archetypes/line_strips2d.fbs";
include "./archetypes/line_strips3d.fbs";
include "./archetypes/mesh3d.fbs";
include "./archetypes/pinhole.fbs";
include "./archetypes/points2d.fbs";
include "./archetypes/points3d.fbs";
include "./archetypes/segmentation_image.fbs";
include "./archetypes/tensor.fbs";
include "./archetypes/text_document.fbs";
include "./archetypes/text_log.fbs";
include "./archetypes/time_series_scalar.fbs";
include "./archetypes/transform3d.fbs";
include "./archetypes/view_coordinates.fbs";

include "./testing/archetypes/fuzzy.fbs";

// NOTE: The current file is the entrypoint for codegen, and so we need to import
// the blueprint types here otherwise nothing will.
include "./blueprint.fbs";

namespace rerun.archetypes;
