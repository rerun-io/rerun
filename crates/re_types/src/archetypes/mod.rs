// DO NOT EDIT! This file was auto-generated by crates/re_types_builder/src/codegen/rust/api.rs

mod annotation_context;
mod arrows3d;
mod bar_chart;
mod boxes2d;
mod boxes2d_ext;
mod boxes3d;
mod boxes3d_ext;
mod clear;
mod clear_ext;
mod depth_image;
mod depth_image_ext;
mod disconnected_space;
mod image;
mod image_ext;
mod line_strips2d;
mod line_strips3d;
mod mesh3d;
mod mesh3d_ext;
mod pinhole;
mod pinhole_ext;
mod points2d;
mod points3d;
mod segmentation_image;
mod segmentation_image_ext;
mod tensor;
mod tensor_ext;
mod text_document;
mod text_log;
mod time_series_scalar;
mod transform3d;
mod view_coordinates;
mod view_coordinates_ext;

pub use self::annotation_context::AnnotationContext;
pub use self::arrows3d::Arrows3D;
pub use self::bar_chart::BarChart;
pub use self::boxes2d::Boxes2D;
pub use self::boxes3d::Boxes3D;
pub use self::clear::Clear;
pub use self::depth_image::DepthImage;
pub use self::disconnected_space::DisconnectedSpace;
pub use self::image::Image;
pub use self::line_strips2d::LineStrips2D;
pub use self::line_strips3d::LineStrips3D;
pub use self::mesh3d::Mesh3D;
pub use self::pinhole::Pinhole;
pub use self::points2d::Points2D;
pub use self::points3d::Points3D;
pub use self::segmentation_image::SegmentationImage;
pub use self::tensor::Tensor;
pub use self::text_document::TextDocument;
pub use self::text_log::TextLog;
pub use self::time_series_scalar::TimeSeriesScalar;
pub use self::transform3d::Transform3D;
pub use self::view_coordinates::ViewCoordinates;
