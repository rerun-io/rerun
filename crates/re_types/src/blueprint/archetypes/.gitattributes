# DO NOT EDIT! This file is generated by crates/re_types_builder/src/lib.rs

.gitattributes linguist-generated=true
mod.rs linguist-generated=true
plot_legend.rs linguist-generated=true
scalar_axis.rs linguist-generated=true
space_view_blueprint.rs linguist-generated=true
space_view_contents.rs linguist-generated=true
