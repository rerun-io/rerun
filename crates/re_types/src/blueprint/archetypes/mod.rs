// DO NOT EDIT! This file was auto-generated by crates/re_types_builder/src/codegen/rust/api.rs

mod background;
mod background_ext;
mod plot_legend;
mod scalar_axis;
mod space_view_blueprint;
mod space_view_contents;
mod visible_time_range;
mod visual_bounds;

pub use self::background::Background;
pub use self::plot_legend::PlotLegend;
pub use self::scalar_axis::ScalarAxis;
pub use self::space_view_blueprint::SpaceViewBlueprint;
pub use self::space_view_contents::SpaceViewContents;
pub use self::visible_time_range::VisibleTimeRange;
pub use self::visual_bounds::VisualBounds;
