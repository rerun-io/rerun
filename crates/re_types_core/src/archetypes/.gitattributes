# DO NOT EDIT! This file is generated by crates/re_types_builder/src/lib.rs

.gitattributes linguist-generated=true
clear.rs linguist-generated=true
mod.rs linguist-generated=true
