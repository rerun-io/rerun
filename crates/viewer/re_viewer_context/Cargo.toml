[package]
authors.workspace = true
description = "Rerun viewer state that is shared with the viewer's code components."
edition.workspace = true
homepage.workspace = true
license.workspace = true
name = "re_viewer_context"
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true
include = ["../../LICENSE-APACHE", "../../LICENSE-MIT", "**/*.rs", "Cargo.toml"]

[lints]
workspace = true

[package.metadata.docs.rs]
all-features = true

[dependencies]
re_chunk_store.workspace = true
re_chunk.workspace = true
re_data_source.workspace = true
re_entity_db = { workspace = true, features = ["serde"] }
re_format.workspace = true
re_log_types.workspace = true
re_log.workspace = true
re_math.workspace = true
re_query.workspace = true
re_renderer = { workspace = true, features = ["serde"] }
re_smart_channel.workspace = true
re_string_interner.workspace = true
re_tracing.workspace = true
re_types = { workspace = true, features = ["ecolor", "glam", "image"] }
re_types_core.workspace = true
re_ui.workspace = true
re_video = { workspace = true, features = ["serde"] }

ahash.workspace = true
anyhow.workspace = true
bit-vec.workspace = true
bitflags.workspace = true
bytemuck.workspace = true
directories.workspace = true
egui_extras.workspace = true
egui_tiles.workspace = true
egui-wgpu.workspace = true
egui.workspace = true
emath.workspace = true
glam = { workspace = true, features = ["serde"] }
half.workspace = true
image = { workspace = true, features = ["jpeg", "png"] }
indexmap = { workspace = true, features = ["std", "serde"] }
itertools.workspace = true
linked-hash-map.workspace = true
ndarray.workspace = true
nohash-hasher.workspace = true
once_cell.workspace = true
parking_lot = { workspace = true, features = ["serde"] }
rfd.workspace = true
serde.workspace = true
slotmap.workspace = true
smallvec.workspace = true
strum_macros.workspace = true
thiserror.workspace = true
uuid = { workspace = true, features = ["serde", "v4", "js"] }
wgpu.workspace = true

# Native dependencies:
[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
arboard = { workspace = true, default-features = false, features = [
  "image-data",
] }

# Web dependencies:
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm-bindgen-futures.workspace = true
