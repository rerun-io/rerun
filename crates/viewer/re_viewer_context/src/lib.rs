//! Rerun Viewer context
//!
//! This crate contains data structures that are shared with most modules of the viewer.

mod annotations;
mod app_options;
mod blueprint_helpers;
mod blueprint_id;
mod cache;
mod collapsed_id;
mod command_sender;
mod component_fallbacks;
mod component_ui_registry;
mod contents;
mod file_dialog;
mod image_info;
mod item;
mod maybe_mut_ref;
mod query_context;
mod query_range;
mod selection_history;
mod selection_state;
mod space_view;
mod store_context;
pub mod store_hub;
mod tensor;
pub mod test_context; //TODO(ab): this should be behind #[cfg(test)], but then ` cargo clippy --all-targets` fails
mod time_control;
mod time_drag_value;
mod typed_entity_collections;
mod utils;
mod viewer_context;

// TODO(andreas): Move to its own crate?
pub mod gpu_bridge;

pub use annotations::{
    AnnotationMap, Annotations, ResolvedAnnotationInfo, ResolvedAnnotationInfos,
};
pub use app_options::AppOptions;
pub use blueprint_helpers::{blueprint_timeline, blueprint_timepoint_for_writes};
pub use blueprint_id::{BlueprintId, BlueprintIdRegistry, ContainerId, SpaceViewId};
pub use cache::{Cache, Caches, ImageDecodeCache, ImageStatsCache, TensorStatsCache, VideoCache};
pub use collapsed_id::{CollapseItem, CollapseScope, CollapsedId};
pub use command_sender::{
    command_channel, CommandReceiver, CommandSender, SystemCommand, SystemCommandSender,
};
pub use component_fallbacks::{
    ComponentFallbackError, ComponentFallbackProvider, ComponentFallbackProviderResult,
    TypedComponentFallbackProvider,
};
pub use component_ui_registry::{ComponentUiRegistry, ComponentUiTypes, UiLayout};
pub use contents::{blueprint_id_to_tile_id, Contents, ContentsName};
pub use image_info::ImageInfo;
pub use item::Item;
pub use maybe_mut_ref::MaybeMutRef;
pub use query_context::{
    DataQueryResult, DataResultHandle, DataResultNode, DataResultTree, QueryContext,
};
pub use query_range::QueryRange;
pub use selection_history::SelectionHistory;
pub use selection_state::{
    ApplicationSelectionState, HoverHighlight, InteractionHighlight, ItemCollection,
    ItemSpaceContext, SelectionHighlight,
};
pub use space_view::{
    DataResult, IdentifiedViewSystem, OptionalSpaceViewEntityHighlight, OverridePath,
    PerSystemDataResults, PerSystemEntities, PropertyOverrides, RecommendedSpaceView,
    SmallVisualizerSet, SpaceViewClass, SpaceViewClassExt, SpaceViewClassLayoutPriority,
    SpaceViewClassRegistry, SpaceViewClassRegistryError, SpaceViewEntityHighlight,
    SpaceViewHighlights, SpaceViewOutlineMasks, SpaceViewSpawnHeuristics, SpaceViewState,
    SpaceViewStateExt, SpaceViewSystemExecutionError, SpaceViewSystemRegistrator,
    SystemExecutionOutput, ViewContext, ViewContextCollection, ViewContextSystem, ViewQuery,
    ViewStates, ViewSystemIdentifier, VisualizableFilterContext,
    VisualizerAdditionalApplicabilityFilter, VisualizerCollection, VisualizerQueryInfo,
    VisualizerSystem,
};
pub use store_context::StoreContext;
pub use store_hub::StoreHub;
pub use tensor::{ImageStats, TensorStats};
pub use time_control::{Looping, PlayState, TimeControl, TimeView};
pub use time_drag_value::TimeDragValue;
pub use typed_entity_collections::{
    ApplicableEntities, IndicatedEntities, PerVisualizer, VisualizableEntities,
};
pub use utils::{auto_color_egui, auto_color_for_entity_path, level_to_rich_text};
pub use viewer_context::{RecordingConfig, ViewerContext};

#[cfg(not(target_arch = "wasm32"))]
mod clipboard;

#[cfg(not(target_arch = "wasm32"))]
pub use clipboard::Clipboard;

pub mod external {
    pub use nohash_hasher;
    pub use {re_chunk_store, re_entity_db, re_log_types, re_query, re_ui};
}

// ---------------------------------------------------------------------------

#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum NeedsRepaint {
    Yes,
    No,
}

// ---

/// Determines the icon to use for a given container kind.
#[inline]
pub fn icon_for_container_kind(kind: &egui_tiles::ContainerKind) -> &'static re_ui::Icon {
    match kind {
        egui_tiles::ContainerKind::Tabs => &re_ui::icons::CONTAINER_TABS,
        egui_tiles::ContainerKind::Horizontal => &re_ui::icons::CONTAINER_HORIZONTAL,
        egui_tiles::ContainerKind::Vertical => &re_ui::icons::CONTAINER_VERTICAL,
        egui_tiles::ContainerKind::Grid => &re_ui::icons::CONTAINER_GRID,
    }
}

/// The style to use for displaying this space view name in the UI.
pub fn contents_name_style(name: &ContentsName) -> re_ui::LabelStyle {
    match name {
        ContentsName::Named(_) => re_ui::LabelStyle::Normal,
        ContentsName::Placeholder(_) => re_ui::LabelStyle::Unnamed,
    }
}
