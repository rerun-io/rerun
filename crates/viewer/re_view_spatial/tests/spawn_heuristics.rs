use re_log_types::Timeline;
use re_types::archetypes;
use re_viewer_context::{test_context::TestContext, ViewClass as _};

use ndarray::{Array, ShapeBuilder as _};

// Creates a 2D scene that mimics the `detect_and_track_objects` example.
#[allow(clippy::unwrap_used)]
fn build_2d_scene(test_context: &mut TestContext) {
    let timeline_step = Timeline::new_sequence("step");
    let time = [(timeline_step, 1)];

    const IMAGE_WIDTH: usize = 64;
    const IMAGE_HEIGHT: usize = 48;

    test_context.log_entity("segmentation/rgb_scaled".into(), |builder| {
        builder.with_archetype(
            re_types::RowId::new(),
            time,
            &archetypes::Image::from_color_model_and_tensor(
                re_types::datatypes::ColorModel::RGB,
                Array::<u8, _>::zeros((IMAGE_HEIGHT, IMAGE_WIDTH, 3).f()),
            )
            .unwrap(),
        )
    });
    test_context.log_entity("segmentation".into(), |builder| {
        builder.with_archetype(
            re_types::RowId::new(),
            time,
            &archetypes::SegmentationImage::try_from(Array::<u8, _>::zeros(
                (IMAGE_HEIGHT, IMAGE_WIDTH).f(),
            ))
            .unwrap(),
        )
    });
    test_context.log_entity("segmentation/detections/things".into(), |builder| {
        builder.with_archetype(
            re_types::RowId::new(),
            time,
            &archetypes::Boxes2D::from_centers_and_half_sizes(
                [(5.0, 5.0), (10.0, 10.0), (15.0, 15.0)],
                [(10.0, 10.0), (10.0, 10.0), (10.0, 10.0)],
            ),
        )
    });

    test_context.log_entity("video".into(), |builder| {
        builder
            .with_archetype(
                re_types::RowId::new(),
                time,
                &archetypes::AssetVideo::from_file_path("../../../tests/assets/empty.mp4").unwrap(),
            )
            .with_archetype(
                re_types::RowId::new(),
                time,
                &archetypes::VideoFrameReference::new(0),
            )
    });
    for i in 0..3 {
        test_context.log_entity(format!("video/tracked/{i}").into(), |builder| {
            builder.with_archetype(
                re_types::RowId::new(),
                time,
                &archetypes::Boxes2D::from_centers_and_half_sizes(
                    [(i as f32, i as f32)],
                    [(1.0, 1.0)],
                ),
            )
        });
    }

    // Since we haven't registered the text view, it won't show up in automatically generated views at all.
    // This is just here to mimic an entity the 2D spatial view can't handle at all.
    test_context.log_entity("description".into(), |builder| {
        builder.with_archetype(
            re_types::RowId::new(),
            re_log_types::TimePoint::default(),
            &archetypes::TextDocument::new("test document"),
        )
    });
}

#[test]
fn test_spatial_view_2d_spawn_heuristics() {
    let mut test_context = TestContext::default();
    // It's important to first register the view class before adding any entities,
    // otherwise the `VisualizerEntitySubscriber` for our visualizers doesn't exist yet,
    // and thus will not find anything applicable to the visualizer.
    test_context.register_view_class::<re_view_spatial::SpatialView2D>();

    build_2d_scene(&mut test_context);

    let view_class = test_context
        .view_class_registry
        .get_class_or_log_error(re_view_spatial::SpatialView2D::identifier());

    test_context.run_in_egui_central_panel(|ctx, _ui| {
        let recommended_views =
            view_class.spawn_heuristics(ctx, &re_log_types::ResolvedEntityPathFilter::properties());

        insta::assert_debug_snapshot!(
            "detect_and_track_objects_like_scene_2d_view_heuristic",
            recommended_views
        );
    });
}
