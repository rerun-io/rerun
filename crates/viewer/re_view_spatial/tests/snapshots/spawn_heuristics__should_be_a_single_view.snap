---
source: crates/viewer/re_view_spatial/tests/spawn_heuristics.rs
expression: recommended_views
---
ViewSpawnHeuristics {
    recommended_views: [
        RecommendedView {
            origin: /3D/camera,
            query_filter: [
                "+ $origin/**",
            ],
        },
        RecommendedView {
            origin: /image,
            query_filter: [
                "+ $origin",
            ],
        },
        RecommendedView {
            origin: /image/color,
            query_filter: [
                "+ $origin/**",
            ],
        },
        RecommendedView {
            origin: /image/depth,
            query_filter: [
                "+ $origin/**",
            ],
        },
        RecommendedView {
            origin: /image/seg,
            query_filter: [
                "+ $origin/**",
            ],
        },
    ],
}
